### PR TITLE
Updated Whiskers for checking invalid tags

### LINKED ISSUE #13226: Strengthen Whiskers' matching
Whiskers' regex looks for `<..>` as opposed to tokenizing the input, so it is not able to detect truncated/mistyped input.

For example it can't detect that `<!b` is not terminated in`<?b> X <!b Y </b>`.

With some heuristics it could be improved: match for `<?`, `<!` and `</` and only allow identifiers which are terminated with `>`, otherwise it is invalid.

_Originally posted by @axic in https://github.com/ethereum/solidity/pull/13219#discussion_r907476574_

### PR DESCRIPTION
**Pull request by rajgaur98**
For Example, raise an error if it contains invalid tags like <?ba<!b</b

Closes #13226

### CODE REVIEW

File: test/libsolutil/Whiskers.cpp [hunk 1]
```diff
@@ -88,8 +88,13 @@ BOOST_AUTO_TEST_CASE(conditional_with_else)
...
 	string templ = "<?b>X<!bY</b>";
-	BOOST_CHECK_EQUAL(Whiskers(templ)("b", true).render(), "X<!bY");
-	BOOST_CHECK_EQUAL(Whiskers(templ)("b", false).render(), "");
+	BOOST_CHECK_THROW(Whiskers _(templ), WhiskersError);
+
+	templ = "<?bX<!b>Y</b>";
+	BOOST_CHECK_THROW(Whiskers _(templ), WhiskersError);
+
+	templ = "<?b>X<!b>Y</b";
+	BOOST_CHECK_THROW(Whiskers _(templ), WhiskersError);
```
Comment (leonardoalt) [inline test/libsolutil/Whiskers.cpp]:
```suggestion
	BOOST_CHECK_THROW(Whiskers{templ}, WhiskersError);
```
how about this?

File: test/libsolutil/Whiskers.cpp [hunk 2]
```diff
@@ -88,8 +88,13 @@ BOOST_AUTO_TEST_CASE(conditional_with_else)
...
 	string templ = "<?b>X<!bY</b>";
-	BOOST_CHECK_EQUAL(Whiskers(templ)("b", true).render(), "X<!bY");
-	BOOST_CHECK_EQUAL(Whiskers(templ)("b", false).render(), "");
+	BOOST_CHECK_THROW(Whiskers{templ}, WhiskersError);
+
+	templ = "<?bX<!b>Y</b>";
+	BOOST_CHECK_THROW(Whiskers{templ}, WhiskersError);
+
+	templ = "<?b>X<!b>Y</b";
+	BOOST_CHECK_THROW(Whiskers{templ}, WhiskersError);
```
Comment (axic) [inline test/libsolutil/Whiskers.cpp:97]: Would be nice if we could inspect the exception message, but didn't find such a feature in boost after a quick look.

**Review by axic — DISMISSED** [review]
Looks good to me.

File: libsolutil/Whiskers.cpp [hunk 3]
```diff
@@ -74,6 +75,17 @@ string Whiskers::render() const
 	return replace(m_template, m_parameters, m_conditions, m_listParameters);
 }
 
+void Whiskers::checkTemplateValid() const
+{
+	regex validTemplate("<[?!\\/][a-zA-Z0-9_$-]+(?:[^a-zA-Z0-9_$>-]|$)");
```
Comment (axic) [inline libsolutil/Whiskers.cpp]: > 	/// Sets a list parameter, <#listName> </listName>.

There are also list parameters which could be truncated. Perhaps should support them here.

Comment (axic) [inline libsolutil/Whiskers.cpp]: While it looks ugly, probably should use `paramRegex()` for identifier part, as that's what other regexes in this file do.

Maybe this needs a refactoring in a separate PR, let's keep it as-is but leave a comment that it must match `paramRegex`.

Comment (rajgaur98) [reply libsolutil/Whiskers.cpp]: Sure, I will include the check for list params

Comment (rajgaur98) [reply libsolutil/Whiskers.cpp]: Ahh I see, totally forgot about `paramRegex`. I can refactor this in another PR

**Review by axic — APPROVED** [review]

**Pull request merged**